diff --git a/CudaSoliditySHA3Solver/CMakeLists.txt b/CudaSoliditySHA3Solver/CMakeLists.txt
index 08d6f82..0550913 100644
--- a/CudaSoliditySHA3Solver/CMakeLists.txt
+++ b/CudaSoliditySHA3Solver/CMakeLists.txt
@@ -43,12 +43,14 @@ endif()
 set(CUDA_NVCC_FLAGS
   ${CUDA_NVCC_FLAGS}
   -std=c++11
-  -Xptxas=-v,-warn-spills
-  -gencode arch=compute_30,code=sm_30
-  -gencode arch=compute_35,code=sm_35
+  -Xptxas=-O3,-v,-warn-spills
+  -Xcompiler=-O3
+  -use_fast_math
   -gencode arch=compute_50,code=sm_50
   -gencode arch=compute_52,code=sm_52
   -gencode arch=compute_61,code=sm_61
+  -gencode arch=compute_86,code=sm_86
+  -gencode arch=compute_89,code=sm_89
 )
 
 if(NOT MSVC)
diff --git a/CudaSoliditySHA3Solver/cudaSha3.cu b/CudaSoliditySHA3Solver/cudaSha3.cu
index 98c6175..f30b060 100644
--- a/CudaSoliditySHA3Solver/cudaSha3.cu
+++ b/CudaSoliditySHA3Solver/cudaSha3.cu
@@ -131,7 +131,7 @@ __device__ __forceinline__ nonce_t rotr(nonce_t input, uint32_t const offset)
 	return input;
 }
 
-__global__ void hashMidstate(uint64_t *__restrict__ solutions, uint32_t *__restrict__ solutionCount, uint32_t maxSolutionCount, uint64_t startPosition)
+__launch_bounds__(1024) __global__ void hashMidstate(uint64_t *__restrict__ solutions, uint32_t *__restrict__ solutionCount, uint32_t maxSolutionCount, uint64_t startPosition)
 {
 	nonce_t nonce, state[25], C[5], D[5], n[11];
 	nonce.uint64 = blockDim.x * blockIdx.x + threadIdx.x + startPosition;
@@ -356,4 +356,4 @@ namespace CUDASolver
 		hashMidstate<<<device->Grid, device->Block>>>(device->SolutionsDevice, device->SolutionCountDevice, device->MaxSolutionCount, device->WorkPosition);
 		CudaSyncAndCheckError(errorMessage);
 	}
-}
\ No newline at end of file
+}
diff --git a/OpenCLSoliditySHA3Solver/CMakeLists.txt b/OpenCLSoliditySHA3Solver/CMakeLists.txt
index e133115..8197120 100644
--- a/OpenCLSoliditySHA3Solver/CMakeLists.txt
+++ b/OpenCLSoliditySHA3Solver/CMakeLists.txt
@@ -88,6 +88,7 @@ file(GLOB SRC_FILES
 )
 
 include_directories($ENV{AMDAPPSDKROOT}/include)
+INCLUDE_DIRECTORIES(${OpenCL_INCLUDE_DIR})
 
 add_library(${PROJECT_NAME} SHARED ${SRC_FILES})
 
diff --git a/OpenCLSoliditySHA3Solver/openCLSolver.cpp b/OpenCLSoliditySHA3Solver/openCLSolver.cpp
index 19d4c72..3a1d37c 100644
--- a/OpenCLSoliditySHA3Solver/openCLSolver.cpp
+++ b/OpenCLSoliditySHA3Solver/openCLSolver.cpp
@@ -451,4 +451,4 @@ namespace OpenCLSolver
 			, "Error setting solution count buffer to kernel"))
 			return;
 	}
-}
\ No newline at end of file
+}
diff --git a/SoliditySHA3Miner/Miner/MinerBase.cs b/SoliditySHA3Miner/Miner/MinerBase.cs
index a7aa289..b240a1b 100644
--- a/SoliditySHA3Miner/Miner/MinerBase.cs
+++ b/SoliditySHA3Miner/Miner/MinerBase.cs
@@ -228,12 +228,12 @@ namespace SoliditySHA3Miner.Miner
 
                 case "DEBUG":
                 default:
-#if DEBUG
+// #if DEBUG
                     sFormat.Append(deviceEnum > -1 ? "[DEBUG] {1}" : "[DEBUG] {0}");
                     break;
-#else
+// #else
                     return;
-#endif
+// #endif
             }
             Program.Print(deviceEnum > -1
                 ? string.Format(sFormat.ToString(), deviceEnum, message)
@@ -484,4 +484,4 @@ namespace SoliditySHA3Miner.Miner
             });
         }
     }
-}
\ No newline at end of file
+}
diff --git a/SoliditySHA3Miner/SoliditySHA3Miner.csproj b/SoliditySHA3Miner/SoliditySHA3Miner.csproj
index 8615c10..871fd49 100644
--- a/SoliditySHA3Miner/SoliditySHA3Miner.csproj
+++ b/SoliditySHA3Miner/SoliditySHA3Miner.csproj
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>netcoreapp2.2</TargetFramework>
+    <TargetFramework>net5.0</TargetFramework>
     <Version>2.2.2</Version>
     <Authors>lwYeo</Authors>
     <Company>lwYeo@github</Company>
